ls1intum/Artemis
# Fix an issue with URL parsing in LTI and improve tests with time mocking

Fix an issue with URL handing in LTI Resource based on the following bug report

```
It seems the ? query parameter is missing after redirect.
U4I Production URL (v4.10.3): https://artemis.university4industry.com/courses/2/exercises/2?welcome=&jwt=eyJhbGciOi
U4I Development URL (v4.12.6):  https://dev-artemis.university4industry.com/courses/2/exercises/2jwt=eyJhbGci...
 
The exercise also seems to be found in the logs before the redirect.
2021-06-18 08:23:13.990  INFO 1 --- [nio-8080-exec-6] d.t.i.www1.artemis.web.rest.LtiResource  : found exercise Adapter Pattern
2021-06-18 08:23:14.105  INFO 1 --- [nio-8080-exec-6] d.t.i.www1.artemis.web.rest.LtiResource  : handleLaunchRequest done
2021-06-18 08:23:14.118  INFO 1 --- [nio-8080-exec-6] d.t.i.www1.artemis.web.rest.LtiResource  : isNewUser: false
2021-06-18 08:23:14.131  INFO 1 --- [nio-8080-exec-6] d.t.i.www1.artemis.web.rest.LtiResource  : created jwt token: eyJhbGciOi…
2021-06-18 08:23:14.144  INFO 1 --- [nio-8080-exec-6] d.t.i.www1.artemis.web.rest.LtiResource  : redirect to url: https://dev-artemis.university4industry.com/courses/2/exercises/2jwt=eyJhbGci...
2021-06-18 08:23:14.966  INFO 1 --- [taskScheduler-2] d.t.in.www1.artemis.config.MetricsBean   : Currently connect users 1 with active websocket subscriptions: 20
2021-06-18 08:23:14.970  INFO 1 --- [MessageBroker-2] o.s.w.s.c.WebSocketMessageBrokerStats    : WebSocketSession[2 current WS(2)-HttpStream(0)-HttpPoll(0), 4 total, 0 closed abnormally (0 connect failure, 0 send limit, 1 transport error)], stompSubProtocol[null], stompBrokerRelay[null], inboundChannel[pool size = 4, active threads = 0, queued tasks = 0, completed tasks = 105], outboundChannel[pool size = 4, active threads = 0, queued tasks = 0, completed tasks = 6], sockJsScheduler[pool size = 2, active threads = 1, queued tasks = 1, completed tasks = 18]
```

### Checklist

### Motivation and Context
I use the proper `UriComponentsBuilder` now which we should use in all places to create the URL instead of custom and magic string handling. I also improved the tests using it and added the possibility to mock the time with a simple TimeService

### Steps for Testing
1. Participate in a MOOC on edX
2. Start an exercise using LTI with an existing user
3. Make sure everything (in particular the URL works correctly)
